cpp/include/raft/neighbors/detail/cagra/add_nodes.cuh

```diff
@@ -0,0 +1,291 @@
+{
+  const std::uint32_t block_size = 256;
+  const std::uint32_t grid_size  = 512;
+
+  // Initizalize counter
+  RAFT_CUDA_TRY(cudaMemsetAsync(count_ptr, 0, sizeof(CounterT) * dataset_size, cuda_stream));
+
+  count_incoming_edges_kernel<<<grid_size, block_size, 0, cuda_stream>>>(
+    count_ptr, graph_ptr, num_entries);
+}
+
+template <class T, class IdxT>
+void add_node_core(raft::resources const& handle,
+                   const raft::neighbors::cagra::index<T, IdxT>& idx,
+                   const raft::device_matrix_view<const T, std::int64_t> additional_dataset_view,
+                   raft::host_matrix_view<IdxT, std::int64_t> updated_graph)
+{
+  using DistanceT                 = float;
+  const std::size_t degree        = idx.graph_degree();
+  const std::size_t dim           = idx.dim();
+  const std::size_t old_size      = idx.dataset().extent(0);
+  const std::size_t num_add       = additional_dataset_view.extent(0);
+  const std::size_t new_size      = old_size + num_add;
+  const std::uint32_t base_degree = degree * 2;
+
+  // Step 0: Calculate the number of incoming edges for each node
+  auto dev_num_incoming_edges =
+    raft::make_device_vector<std::uint32_t, std::uint64_t>(handle, new_size);
+  auto host_num_incoming_edges = raft::make_host_vector<std::uint32_t, std::uint64_t>(new_size);
+  count_incoming_edges(dev_num_incoming_edges.data_handle(),
+                       idx.graph().data_handle(),
+                       degree * old_size,
+                       new_size,
+                       raft::resource::get_cuda_stream(handle));
+  raft::copy(host_num_incoming_edges.data_handle(),
+             dev_num_incoming_edges.data_handle(),
+             new_size,
+             raft::resource::get_cuda_stream(handle));
+
+  const std::size_t max_chunk_size = 1024;
+
+  raft::neighbors::cagra::search_params params;
+  params.itopk_size = std::max(base_degree * 2lu, 256lu);
+
+  // Memory space for rank-based neighbor list
+  auto neighbor_indices =
+    raft::make_device_matrix<IdxT, std::int64_t>(handle, max_chunk_size, base_degree);
+
+  auto neighbor_distances =
+    raft::make_device_matrix<DistanceT, std::int64_t>(handle, max_chunk_size, base_degree);
+
+  auto queries = raft::make_device_matrix<T, std::int64_t>(handle, max_chunk_size, dim);
+
+  auto host_neighbor_indices =
+    raft::make_host_matrix<IdxT, std::int64_t>(max_chunk_size, base_degree);
+
+  auto neighbors_vectors =
+    raft::make_device_matrix<T, std::int64_t>(handle, max_chunk_size * base_degree, dim);
+
+  auto two_hop_neighbors_indices =
+    raft::make_device_matrix<IdxT, std::int64_t>(handle, max_chunk_size * base_degree, base_degree);
+
+  auto host_two_hop_neighbors_indices =
+    raft::make_host_matrix<IdxT, std::int64_t>(max_chunk_size * base_degree, base_degree);
+
+  auto two_hop_neighbors_distances = raft::make_device_matrix<DistanceT, std::int64_t>(
+    handle, max_chunk_size * base_degree, base_degree);
```

Comment:
We should  probably use the workspace allocator for host allocations. Example:
https://github.com/rapidsai/raft/blob/9fb05a2ab3d72760a09f1b7051e711d773682ef1/cpp/include/raft/neighbors/detail/ivf_pq_build.cuh#L1256-L1259

Reply:
I fixed the code to use the workspace allocator. Is there also a workspace allocator?